### PR TITLE
add gripper component

### PR DESCRIPTION
**Pull request by fotnitfpro000**

### CODE REVIEW

File: tests/vision.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+mjpegServer.setSource(camera)
+
+print("mjpg server listening at http://0.0.0.0:8081")
+input("Press enter to exit...")
+
+frame = numpy.zeros(shape=(480, 640, 3), dtype=numpy.uint8)
+cvSink = CS.getVideo()
+
+
+while True:
```
Comment (LucienMorey) [inline tests/vision.py:21]: This file should not be in this commit

File: tests/vision1.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#rec = 1
+#mask = cv2.inRange()
+
+img = cv2.imread(R"C:\Users\amber\OneDrive\Pictures\xpicture.jpg",cv2.IMREAD_ANYCOLOR)
+frame = numpy.zeros(shape=(720, 1280, 3), dtype=numpy.uint8)
+520,43,665,220
+start_point = (520,43) #top left corner of rectangle
+end_point = (665,220) #bottom right corner of rectangle
+
+while True:
```
Comment (LucienMorey) [inline tests/vision1.py:15]: This file should not be in this commit

File: robot.py [hunk 3]
```diff
@@ -22,6 +24,7 @@ class MyRobot(magicbot.MagicRobot):
 
     # Components
     chassis: Chassis
+    claw: Claw
 
     def createObjects(self) -> None:
         self.logger.info("pyrapidreact %s", GIT_INFO)
```
Comment (LucienMorey) [inline robot.py:30]: You will need to create an instance of the claw solenoid and make reference to the ids for it

File: components/claw.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        is_gripper_open = True
+
+    def close_gripper(self):
+        is_gripper_open = False
+
+    def engage(self):
+        if self.is_gripper_open:
+            self.gripper_solenoid.set(value=wpilib.DoubleSolenoid.Value.kForward)
+        else:
+            self.gripper_solenoid.set(value=wpilib.DoubleSolenoid.Value.kReverse)
```
Comment (LucienMorey) [inline components/claw.py:21]:
```suggestion
    def execute(self):
        if self.is_gripper_open:
            self.gripper_solenoid.set(value=wpilib.DoubleSolenoid.Value.kForward)
        else:
            self.gripper_solenoid.set(value=wpilib.DoubleSolenoid.Value.kReverse)
```

File: components/claw.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import wpilib
+
+
+class Claw:
+    def __init__(self):
+        self.gripper_solenoid = wpilib.DoubleSolenoid(
+            wpilib.PneumaticsModuleType.CTREPCM, forwardChannel=0, reverseChannel=1
+        )
+        self.is_gripper_open = True
```
Comment (LucienMorey) [inline components/claw.py:9]:
```suggestion
    self.gripper_solenoid: wpilib.DoubleSolenoid
    def __init__(self):
        self.is_gripper_open = True
```

File: components/claw.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import wpilib
```
Comment (LucienMorey) [inline components/claw.py:1]: You could also consider importing just the things you need. ie ```DoubleSolenoid```

**Review by LucienMorey — CHANGES_REQUESTED** [review]

**Review by DensityInfinite — APPROVED** [review]

File: components/gripper.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import wpilib
+
+
+class Gripper:
+    gripper_solenoid: wpilib.DoubleSolenoid
+    gripper_game_piece_switch: wpilib.DigitalInput
```
Comment (DensityInfinite) [inline components/gripper.py:6]: Should those be the ones from robot.py?

File: components/gripper.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        self.opened = True
+
+    def close(self):
+        self.opened = False
+
+    def execute(self):
+        if self.opened:
+            self.solenoid.set(value=DoubleSolenoid.Value.kForward)
+        else:
+            self.solenoid.set(value=DoubleSolenoid.Value.kReverse)
```
Comment (auscompgeek) [inline components/gripper.py:30]: nit: it's already obvious we're setting the value
```suggestion
            self.solenoid.set(DoubleSolenoid.Value.kForward)
        else:
            self.solenoid.set(DoubleSolenoid.Value.kReverse)
```

File: ids.py [hunk 9]
```diff
@@ -35,11 +35,20 @@ class Intake:
...
+    class Gripper:
+        gripper_solenoid_forward = 0
+        gripper_solenoid_reverse = 1
+
 
 class PwmChannels:
     ...
 
 
+class DioChannels:
```
Comment (auscompgeek) [inline ids.py:47]: I think this needs to be added to the test

File: components/gripper.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
+from wpilib import DoubleSolenoid, DigitalInput, PneumaticsModuleType
+
+import ids
+
+
+class Gripper:
+    def __init__(self):
```
Comment (auscompgeek) [inline components/gripper.py:7]: Ensure all the methods are covered by the type checker.

```suggestion
    def __init__(self) -> None:
```

**Review by auscompgeek — APPROVED** [review]
LGTM with comments

**Pull request merged**